Implement gateway S3 support

nl5887 opened the pull request:
Implement gateway s3 backend.

* the s3 gateway layer will connect to minio or s3 endpoints

nl5887 committed:
Implement first version of S3 and caching layer

* the s3 layer will work with Amazon S3 and Minio
* the cache layer will cache GetObject objects locally

nl5887 committed:
Implement first version of gateway caching and s3 support

* the s3 gateway layer will connect to minio or s3 endpoints
* the caching layer will cache objects for all backend layers

This is a very basic and first version. Needs thoroughly testing and
the caching needs all kind of functionality as locking.

nl5887 committed:
Fix linting errors

nl5887 committed:
Fix some more linting errors

nl5887 committed:
Add ErrBucketAlreadyOwnedByYou object error

nl5887 committed:
Fix Mint s3 gateway errors

krishnasrinivas commented on cmd/gateway-main.go:
```diff
@@ -55,7 +55,26 @@ var gatewayCmd = cli.Command{
-	}),
+	},
+		cli.StringFlag{
+			Name:  "s3-endpoint",
```
`endpoint` is better than `s3-endpoint`

krishnasrinivas commented on cmd/gateway-main.go:
```diff
@@ -55,7 +55,26 @@ var gatewayCmd = cli.Command{
+			Value: "play.minio.io:9000",
+		},
+		cli.StringFlag{
+			Name:  "s3-location",
```
`location` is not needed as minio-go figures it out automatically

krishnasrinivas commented on cmd/gateway-main.go:
```diff
@@ -83,11 +103,14 @@ func mustGetGatewayCredsFromEnv() (accessKey, secretKey string) {
-func newGatewayLayer(backendType, accessKey, secretKey string) (GatewayLayer, error) {
-	if gatewayBackend(backendType) != azureBackend {
+func newGatewayLayer(backendType, endpoint, location, accessKey, secretKey string) (GatewayLayer, error) {
+	if gatewayBackend(backendType) == azureBackend {
```
switch statement will be better

krishnasrinivas commented on cmd/gateway-main.go:
```diff
@@ -55,7 +55,26 @@ var gatewayCmd = cli.Command{
+		cli.StringFlag{
+			Name:  "s3-endpoint",
+			Usage: "The endpoint.",
+			Value: "play.minio.io:9000",
```
default should be `s3.amazonaws.com`

krishnasrinivas commented on cmd/s3-layer-anonymous.go:
```diff
@@ -0,0 +1,34 @@
+// AnonGetObject - Get object anonymously
+func (l *S3Layer) AnonGetObject(bucket string, key string, startOffset int64, length int64, writer io.Writer) (err error) {
+	// already an s3 object error
+	return l.GetObject(bucket, key, startOffset, length, writer)
```
it's doing an authenticated operation here

krishnasrinivas commented on cmd/s3-layer-anonymous.go:
```diff
@@ -0,0 +1,34 @@
+// AnonGetObjectInfo - Get object info anonymously
+func (l *S3Layer) AnonGetObjectInfo(bucket string, object string) (objInfo ObjectInfo, err error) {
+	// already an s3 object error
+	return l.GetObjectInfo(bucket, object)
```
it's doing an authenticated operation here

krishnasrinivas commented on cmd/s3-layer-anonymous.go:
```diff
@@ -0,0 +1,34 @@
+
+// AnonListObjects - List objects anonymously
+func (l *S3Layer) AnonListObjects(bucket string, prefix string, marker string, delimiter string, maxKeys int) (result ListObjectsInfo, err error) {
+	result, err = l.ListObjects(bucket, prefix, marker, delimiter, maxKeys)
```
it's doing an authenticated operation here

krishnasrinivas commented on cmd/s3-layer-anonymous.go:
```diff
@@ -0,0 +1,34 @@
+
+// AnonGetBucketInfo - Get bucket metadata anonymously.
+func (l *S3Layer) AnonGetBucketInfo(bucket string) (bucketInfo BucketInfo, err error) {
+	bucketInfo, err = l.GetBucketInfo(bucket)
```
it's doing an authenticated operation here

krishnasrinivas commented on cmd/s3-layer.go:
```diff
@@ -0,0 +1,508 @@
+		}
+	case "XMinioInvalidObjectName":
+		err = ObjectNameInvalid{}
+	default:
```
default statement (and hence fmt.Println statement) can be removed

krishnasrinivas commented on cmd/s3-layer.go:
```diff
@@ -0,0 +1,508 @@
+	s3LayerConfig
+
+	Client *minio.Core
+}
```
s3Layer can be renamed to s3Gateway because we refer to http handlers as s3Layer during discussions. It can contain just two fields:

```
type struct s3Gateway {
    client *minio.Core                   // For authenticated requests
    anonClient *minio.Core          // For unauthenticated requests
}
```

There is no need to have the struct `s3LayerConfig`

krishnasrinivas commented on cmd/s3-layer.go:
```diff
@@ -0,0 +1,508 @@
+	Client *minio.Core
+}
+
+func newS3Layer(endpoint, location, accessKey, secretKey string) (GatewayLayer, error) {
```
all functions need comments (even though it's redundant sometimes, we follow this convention)

krishnasrinivas commented on cmd/s3-layer.go:
```diff
@@ -0,0 +1,508 @@
+
+	i := 0
+
+	for message := range l.Client.ListObjectsV2(bucket, prefix, true, doneCh) {
```
you can use https://github.com/minio/minio-go/blob/master/core.go : ListObjects()

krishnasrinivas commented on cmd/s3-layer.go:
```diff
@@ -0,0 +1,508 @@
+
+// CopyObject - Copies a blob from source container to destination container.
+func (l *S3Layer) CopyObject(srcBucket string, srcObject string, destBucket string, destObject string, metadata map[string]string) (ObjectInfo, error) {
+	err := l.Client.CopyObject(destBucket, destObject, fmt.Sprintf("%s/%s", srcBucket, srcObject), minio.CopyConditions{})
```
you can use pathJoin() helper instead of fmt.Sprintf

krishnasrinivas commented on cmd/s3-layer.go:
```diff
@@ -0,0 +1,508 @@
+
+// CopyObjectPart - copy part of object to other bucket and object
+func (l *S3Layer) CopyObjectPart(srcBucket string, srcObject string, destBucket string, destObject string, uploadID string, partID int, startOffset int64, length int64) (info PartInfo, err error) {
+	panic("CopyObjectPart: not implemented")
```
you can remove panic() statements here and at other parts of the code and instead return errors.

krishnasrinivas commented on cmd/s3-layer.go:
```diff
@@ -0,0 +1,508 @@
+}
+
+// SetBucketPolicies - Set policy on bucket
+func (l *S3Layer) SetBucketPolicies(string, []BucketAccessPolicy) error {
```
this should be implemented.

krishnasrinivas commented on cmd/s3-layer.go:
```diff
@@ -0,0 +1,508 @@
+}
+
+// GetBucketPolicies - Get policy on bucket
+func (l *S3Layer) GetBucketPolicies(string) ([]BucketAccessPolicy, error) {
```
this should be implemented.

krishnasrinivas commented on cmd/s3-layer.go:
```diff
@@ -0,0 +1,508 @@
+}
+
+// DeleteBucketPolicies - Delete all policies on bucket
+func (l *S3Layer) DeleteBucketPolicies(string) error {
```
this should be implemented.

nl5887 replied:
Still need to implement the BucketPolicy functions, all other issues have been solved.

krishnasrinivas reviewed (changes requested):
I will review the caching code next.

nl5887 committed:
Fix reviewing issues

- use endpoint instead of s3-endpoint
- change default endpoint to s3.amazonaws.com
- removed location
- use switch instead of 2 if statement
- fixed anon operations
- changed default statement to custom error instead of println
- added anonclient, removed config
- added comments to functions
- use ListObjects of core
- use path.Join instead of fmt.Sprintf
- removed panic, returning NotImplemented instead

krishnasrinivas commented on cmd/s3-layer-anonymous.go:
```diff
@@ -3,32 +3,62 @@ package cmd
+		return s3ToObjectError(traceError(err), bucket, key)
+	}
+
+	object.Close()
```
`defer object.Close()` should be done and moved to the line 11

krishnasrinivas commented on cmd/s3-layer-anonymous.go:
```diff
@@ -3,32 +3,62 @@ package cmd
+
+	result, err := l.anonClient.ListObjects(bucket, prefix, marker, delimiter, maxKeys)
 	if err != nil {
-		return ListObjectsInfo{}, s3ToObjectError(traceError(err), bucket)
```
`return ListObjectsInfo{}, s3ToObjectError(traceError(err), bucket)` should be done?

krishnasrinivas commented on cmd/s3-layer-anonymous.go:
```diff
@@ -3,32 +3,62 @@ package cmd
 // AnonGetBucketInfo - Get bucket metadata anonymously.
-func (l *S3Layer) AnonGetBucketInfo(bucket string) (bucketInfo BucketInfo, err error) {
-	bucketInfo, err = l.GetBucketInfo(bucket)
+func (l *s3Gateway) AnonGetBucketInfo(bucket string) (BucketInfo, error) {
```
Since S3 does not allow anonymous listing of buckets, what you can do is do `l.anonClient.BucketExists()` if there is no error then do authenticated ListBuckets to get the `creationDate`, if there was any error from `l.anonClient.BucketExists()` then we can just return error.

krishnasrinivas commented on cmd/s3-layer.go:
```diff
@@ -77,73 +78,68 @@ func s3ToObjectError(err error, params ...string) error {
 	default:
-		fmt.Println("Undefined s3 object error:", bucket, object, minioErr.Code)
+		// should not be reached
+		err = fmt.Errorf("Undefined s3 object error: bucket=%s object=%s code=%s", bucket, object, minioErr.Code)
```
This will cause the original error to be replaced with fmt.Errorf() error, hence you could just remove the entire `default` section from the switch statement. There is a high chance of us reaching default statement as number `if` cases in the switch statement is limited.

krishnasrinivas commented on cmd/s3-layer.go:
```diff
@@ -77,73 +78,68 @@ func s3ToObjectError(err error, params ...string) error {
+// newS3Gateway returns s3 gatewaylayer
+func newS3Gateway(endpoint, accessKey, secretKey string) (GatewayLayer, error) {
+	// Initialize minio client object.
+	client, err := minio.NewCore(endpoint, accessKey, secretKey, true)
```
since we need to allow for http and https, endpoint would be `http://host.com` so endpoint needs to be parsed to see if the scheme is http or https and accordingly call `minio.NewCore` with the isSecure option

krishnasrinivas commented on cmd/s3-layer.go:
```diff
@@ -191,34 +187,35 @@ func (l *S3Layer) DeleteBucket(bucket string) error {
-		}
+	result, err := l.Client.ListObjects(bucket, prefix, marker, delimiter, maxKeys)
+	if err != nil {
+		return ListObjectsInfo{}, err
```
return s3ToObjectError(traceError(message.Err), bucket) ?

nl5887 committed:
Fix reviewing errors

- moved object.Close
- return s3ToObjectError
- check for BucketExists
- removed default error
- parse url for http / https and endpoint

nl5887 committed:
Merge remote-tracking branch 'origin/master' into feature-gateway-cache

nl5887 committed:
Fix linting error

nl5887 committed:
Fix linting error

nl5887 committed:
Split of caching layer and s3 gateway

nl5887 committed:
Merge branch 'master' into feature-gateway-cache

krisis commented on cmd/s3-layer.go:
```diff
@@ -0,0 +1,511 @@
+
+// MakeBucket - Create a new container on S3 backend.
+func (l *s3Gateway) MakeBucket(bucket string) error {
+	err := l.Client.MakeBucket(bucket, "us-east-1")
```
we should make "us-east-1" a constant.

harshavardhana replied:
should be made configurable.

nl5887 replied:
had this configurable, this is comment of @krishnasrinivas:

```location is not needed as minio-go figures it out automatically```

harshavardhana replied:
It is useful for bucket creates.  Here it defaults to `us-east-1` perhaps even need to configure during newGatewayConfig() for different S3 endpoints. Through region as ENV. We need to discuss @krishnasrinivas 

Otherwise al requests would go to same endpoint us-east.

krishnasrinivas replied:
@harshavardhana yes it's needed for bucket creation, actually we can use the location info sent by the client:
```
// createBucketConfiguration container for bucket configuration.                                                                                                                        
type createBucketConfiguration struct {                                                                                                                                                 
        XMLName  xml.Name `xml:"http://s3.amazonaws.com/doc/2006-03-01/ CreateBucketConfiguration" json:"-"`                                                                            
        Location string   `xml:"LocationConstraint"`                                                                                                                                    
}
```

harshavardhana replied:
This might be sufficient.. for location.

krisis commented on cmd/s3-layer.go:
```diff
@@ -0,0 +1,511 @@
+func (l *s3Gateway) ListObjects(bucket string, prefix string, marker string, delimiter string, maxKeys int) (ListObjectsInfo, error) {
+	loi := ListObjectsInfo{}
+
+	loi.Objects = make([]ObjectInfo, maxKeys)
```
`loi` is unused.

donatello commented on cmd/s3-layer.go:
```diff
@@ -0,0 +1,511 @@
+
+// PutObjectPart puts a part of object in bucket
+func (l *s3Gateway) PutObjectPart(bucket string, object string, uploadID string, partID int, size int64, data io.Reader, md5Hex string, sha256sum string) (PartInfo, error) {
+	md5HexBytes, err := hex.DecodeString(md5Hex)
```
Why is `md5Hex` argument hex encoded here? It is getting decoded here, and then getting re-encoded as base64 again inside minio-go

donatello commented on cmd/s3-layer.go:
```diff
@@ -0,0 +1,511 @@
+		return PartInfo{}, err
+	}
+
+	sha256sumBytes, err := hex.DecodeString(sha256sum)
```
Why is `sha256sum` argument hex encoded here? It is getting decoded here, and then getting re-encoded as base64 again inside minio-go

nl5887 replied:
the minio-go core expects []byte if we want to pass it as base64 string we need to adjust core. In my opinion it is better to pass it as []byte and just settle with the de- and encoding.

krishnasrinivas replied:
At the API level it would be more appropriate to represent md5/sha as `string` ... I will make those changes to minio-go. It will make S3 gateway code simpler too, it would just pass along these values instead of converting from string->[]byte

donatello replied:
Just had a better look at this - sha256sum is used hex encoded and md5sum is used base64 encoded ("used" => S3 API requirement). There is some unneeded/repeated conversion between binary ([]byte) to string (hex/base64 encodings) representations and vice versa. 

Ideally, hashes need an explicit type that store just bits (using []byte) and have methods for conversion to various formats like hex or base64 (and this can be taken up outside this patch).

krishnasrinivas replied:
that is a better way to represent md5 and sha256:
`[md5.Size]byte` and `[sha256.Size]byte`

nl5887 replied:
I'd oppose to changing minio-go to strings, because those are undefined (we don't know if they are valid strings). I'd say you want to pass byte hashes instead of strings. And make the appropriate conversions where necessary, eg in the gateway.

krisis commented on cmd/s3-layer.go:
```diff
@@ -0,0 +1,511 @@
+		return ObjectInfo{}, s3ToObjectError(traceError(err), bucket, object)
+	}
+
+	if sha256sum != "" {
```
minio-go computes sha256 sum, why do we need to compute it at the gateway too?

nl5887 replied:
just tested it, the Mint TestSHA256Mismatch fails without this code. The sha 256 mismatch error isn't being returned by minio-go.

krisis replied:
interesting, this https://github.com/minio/minio-go/blob/master/api-put-object.go#L204 made me believe that for V4 signature and non-HTTPS client, minio-go SDK computes sha256.

nl5887 replied:
yeah, I was looking at the same. It calculates them, but won't return an error at all.

harshavardhana replied:
Server doing Sha is correct we do this with Azure as well. Double checksum is fine. Usually gateway is talking to an SSL enabled remote S3 server in those cases minio-go will not compute sha256. We can document this as a caveat but we storngly recommend using a secure remote S3. 

With Amazon S3 as backend it's already solved.

harshavardhana replied:
Your streaming signature change would help here @krisis as well.

krishnasrinivas replied:
we can expose the following function in minio-go:
```
func (c Client) putObjectDo(bucketName, objectName string, reader io.Reader, md5Sum []byte, sha256Sum []byte, size int64, metaData map[string][]string) (ObjectInfo, error) {
```

harshavardhana replied:
We don't need do it doesn't add any benefit.  A sha should be validated by the gateway not minio-go.  minio-go is just calculating upon incoming stream and uploading.

krishnasrinivas replied:
It adds benefit, i.e the sha256 verification will be done by the backend S3 server not minio-go. Else gateway will have to compute sha256 twice - once in object-layer and once in minio-go.

harshavardhana replied:
Again looking at the code it makes sense to add what you are saying, avoids the RemoveObject() and returns ObjectInfo{} is one call.

nl5887 committed:
Remove unused code

donatello commented on cmd/gateway-main.go:
```diff
@@ -83,11 +91,23 @@ func mustGetGatewayCredsFromEnv() (accessKey, secretKey string) {
+			endpoint = u.Host
+		}
+
+		return newS3Gateway(endpoint, useHTTPS, accessKey, secretKey)
```
```go
case s3Backend:
	if u, err := url.Parse(endpoint); err != nil {
		return nil, err
	}

	return newS3Gateway(u.Host, u.Scheme == "https", accessKey, secretKey)
```

I think concise is better.

nl5887 replied:
Current code allows urls without scheme as well. Eg. both http://play.minio.io:9000 and play.minio.io:9000 are supported now.

harshavardhana replied:
The url.Parse fails for address in format "play.minio.io:9000" in go1.8 , i would suggest we avoid that kind of support or if needed explicitly handle it using `net.SplitHostPort()` 

https://github.com/golang/go/commit/c5ccbdd22bdbdc43d541b7e7d4ed66ceb559030e#comments

nl5887 replied:
We need to try to parse the endpoint anyhow to see get the scheme. I'll remove the err!=nil check to pass and default to use https.

donatello commented on cmd/s3-layer-anonymous.go:
```diff
@@ -0,0 +1,71 @@
+func (l *s3Gateway) AnonListObjects(bucket string, prefix string, marker string, delimiter string, maxKeys int) (ListObjectsInfo, error) {
+	loi := ListObjectsInfo{}
+
+	loi.Objects = make([]ObjectInfo, maxKeys)
```
`loi` is unused.

donatello commented on cmd/s3-layer.go:
```diff
@@ -0,0 +1,511 @@
+}
+
+// GetObjectInfo - reads blob metadata properties and replies back ObjectInfo,
+// uses zure equivalent GetBlobProperties.
```
Outdated comment.

nl5887 committed:
Use location of createBucketConfiguration for MakeBucket

nl5887 committed:
Fix incorrect method description

nl5887 committed:
Merge branch 'master' into feature-gateway-cache

nl5887 committed:
Update new method signature

nl5887 committed:
Update tests to new MakeBucket signature

krishnasrinivas commented on cmd/azure.go:
```diff
@@ -151,7 +151,7 @@ func (a AzureObjects) StorageInfo() StorageInfo {
 
 // MakeBucket - Create a new container on azure backend.
-func (a AzureObjects) MakeBucket(bucket string) error {
+func (a AzureObjects) MakeBucket(location, bucket string) error {
```
you can override MakeBucket in gateway-interface as location does not make sense for fs/xl

nl5887 replied:
That will imply a lot of changes:

* ObjectLayer cannot be implemented anymore within GatewayLayer, because MakeBucket has different signature within ObjectLayer and GatewayLayer and cannot override. So need to copy all methods from interface ObjectLayer to GatewayLayer
* globalObjectAPI cannot be used anymore, because we have ObjectLayer and GatewayLayer difererent signature for MakeBucket, need to find other solution for that.

Now I've chosen to add a MakeBucketWithLocation to comply to GatewayLayer interface and MakeBucket to comply to ObjectLayer interface which will call the default region. The MakeBucket will be never called because PutBucket has been overridden for the GatewayLayer.

nl5887 committed:
Remove unused code

nl5887 committed:
Revert "Update tests to new MakeBucket signature"

This reverts commit f775719d441426ce65d4867097be7f6dfa418056.

nl5887 committed:
Revert "Use location of createBucketConfiguration for MakeBucket"

This reverts commit a23b36034d3b25b0bc8ff991ca41a8a40608db54.

harshavardhana commented on cmd/fs-v1.go:
```diff
@@ -203,7 +203,7 @@ func (fs fsObjects) statBucketDir(bucket string) (os.FileInfo, error) {
 // MakeBucket - create a new bucket, returns if it
 // already exists.
-func (fs fsObjects) MakeBucket(bucket string) error {
+func (fs fsObjects) MakeBucket(location, bucket string) error {
```
Location should be the last argument .. All optional arguments go to the end.   Also this method signature change is not needed.

harshavardhana commented on cmd/handler-utils.go:
```diff
@@ -27,7 +27,7 @@ import (
 // The location value in the request body should match the
 // region configured at serverConfig, otherwise error is returned.
-func isValidLocationConstraint(r *http.Request) (s3Error APIErrorCode) {
+func isValidLocationConstraint(r *http.Request) (location string, s3Error APIErrorCode) {
```
isValidLocationConstraint should never return `location`   change this function to take just the `location` string and parse the location outside in a different function as 

```go
func parseLocationConstraint(r *http.Request)  (location string, s3Error APIErrorCode) {
        // If the request has no body with content-length set to 0,
	// we do not have to validate location constraint. Bucket will
	// be created at default region.
	locationConstraint := createBucketLocationConfiguration{}
	err = xmlDecoder(r.Body, &locationConstraint, r.ContentLength)
        if err != nil && err != io.EOF {
 	       errorIf(err, "Unable to xml decode location constraint")
               return "", ErrMalformedXML
        }
        // Successfully decoded, proceed to verify the region.
	// Once region has been obtained we proceed to verify it.
	location = locationConstraint.Location
	if location == "" {
		// Location constraint is empty for region globalMinioDefaultRegion,
		// in accordance with protocol.
		location = globalMinioDefaultRegion
	}
        return location, nil 
}

func isValidLocation(location string) (s3Error APIErrorCode) {
         if serverConfig.GetRegion() != location {
                  return ErrInvalidRegion
         }
         return ErrNone
}

```

nl5887 committed:
Add location support to MakeBucket

* to comply with ObjectLayer dummy MakeBucket has been added
* MakeBucketWithLocation has been added to GatewayLayer to allow passing
location
* PutBucketHandler has been overridden in the gateway handler to
call MakeBucketWithLocation

nl5887 committed:
Correct misspelling

nl5887 committed:
Continue on url parsing errors, default to https

Url parsing errors will fallback to default url handling of minio go library.

harshavardhana commented on cmd/handler-utils.go:
```diff
@@ -49,11 +47,28 @@ func isValidLocationConstraint(r *http.Request) (s3Error APIErrorCode) {
+// Validates location constraint in PutBucket request body.
+// The location value in the request body should match the
+// region configured at serverConfig, otherwise error is returned.
+func isValidLocationConstraint(r *http.Request) (s3Error APIErrorCode) {
```
This is changed in the master to simplify this code.. you might have to rebase.

harshavardhana commented on cmd/s3-layer-anonymous.go:
```diff
@@ -0,0 +1,67 @@
+package cmd
```
Add license header..

harshavardhana commented on cmd/s3-layer-anonymous.go:
```diff
@@ -0,0 +1,67 @@
+
+// AnonGetBucketInfo - Get bucket metadata anonymously.
+func (l *s3Gateway) AnonGetBucketInfo(bucket string) (BucketInfo, error) {
+	if exists, err := l.anonClient.BucketExists(bucket); err != nil {
```
To optimize this perhaps we don't need to do `ListBuckets()` just to fill in `BucketInfo()` - since this call is used only in HeadBucketHandler() we can simply change the API to not return BucketInfo() at all. 

In all its usage i see that `BucketInfo{}` is not used at all.

krishnasrinivas replied:
> In all its usage i see that BucketInfo{} is not used at all.

meaning?
`mc ls s3/bucketname` issues a HEAD on the bucket

so returning a BucketInfo{} will be necessary ... right?

harshavardhana replied:
No it is not necessary BucketInfo{} is never used HEAD has no body. @krishnasrinivas we only need to see if Bucket indeed exists.. GetBucketInfo() was implemented for XL and FS which implement the Stat() call.  Doing `ListBuckets()` and filtering is not needed..

nl5887 replied:
The function is called GetBucketInfo so we should return a valid BucketInfo with correct fields if we keep GetBucketInfo. So we'd better remove GetBucketInfo for gateway and replace it with BucketExists.

harshavardhana replied:
Agreed we should do that makes more sense in gateway context.

harshavardhana commented on cmd/s3-layer-unsupported.go:
```diff
@@ -0,0 +1,26 @@
+package cmd
```
license header..

harshavardhana commented on cmd/s3-layer.go:
```diff
@@ -0,0 +1,512 @@
+	return e
+}
+
+// s3Gateway - Implements gateway for S3 and Minio blob storage.
```
s/"Implements gateway for Minio and S3 compatible object storage servers."/

harshavardhana commented on cmd/s3-layer.go:
```diff
@@ -0,0 +1,512 @@
+
+// ListObjects - lists all blobs in S3 bucket filtered by prefix
+func (l *s3Gateway) ListObjects(bucket string, prefix string, marker string, delimiter string, maxKeys int) (ListObjectsInfo, error) {
+	result, err := l.Client.ListObjects(bucket, prefix, marker, delimiter, maxKeys)
```
Perhaps we should use the latest V2 ListObjects API. Needs changes in `minio-go` core. // cc @krishnasrinivas

harshavardhana replied:
We should also implement it at gateway handler layer. Because some aws-sdks might start making V2 List default. So we need to bring it in even for Azure gateway. // cc @krishnasrinivas

krishnasrinivas replied:
yes @harshavardhana https://github.com/minio/minio/issues/4043 currently only authenticated list-v2 works.

harshavardhana commented on cmd/s3-layer.go:
```diff
@@ -0,0 +1,512 @@
+	return fromMinioClientObjectInfo(bucket, oi), nil
+}
+
+// PutObject - Create a new blob with the incoming data,
```
Let's call it `blob` it is only of Azure its called `blob`, we should call it object here in S3 context.

harshavardhana commented on cmd/s3-layer.go:
```diff
@@ -0,0 +1,512 @@
+}
+
+// CopyObjectPart - copy part of object to other bucket and object
+func (l *s3Gateway) CopyObjectPart(srcBucket string, srcObject string, destBucket string, destObject string, uploadID string, partID int, startOffset int64, length int64) (info PartInfo, err error) {
```
Perhaps add a FIXME here, so that we can open relevant bugs and address this.

harshavardhana commented on cmd/s3-layer.go:
```diff
@@ -0,0 +1,512 @@
+
+// DeleteBucketPolicies - Delete all policies on bucket
+func (l *s3Gateway) DeleteBucketPolicies(string) error {
+	return traceError(NotImplemented{})
```
We need to implement this..

nl5887 replied:
Don't you think the policies are too backend specific (eg if we're going to add more and more backends, with all their specifics? I assume people using the gateway will use it for main functionality, not for configuration. They'll use the backend directly.)

krishnasrinivas replied:
@nl5887 we'll only support a subset of {private, read-only, read-write, write-only} depending on what the backend supports.
for example for azure we support {private, read-only}

harshavardhana commented on cmd/s3-layer_test.go:
```diff
@@ -0,0 +1 @@
+package cmd
```
License header missing.

nl5887 committed:
Add license headers

nl5887 committed:
Merge branch 'master' into feature-gateway-cache

nl5887 committed:
Fix comment

nl5887 committed:
Fix comment

nl5887 committed:
Add FIXME

nl5887 committed:
Bump minio-go version

nl5887 committed:
Add s3 gateway support for ListObjectsV2

nl5887 committed:
Implement bucket policy functions

* Mint doesn't succeed (yet), because policies aren't combined
which Mint expects to be combined
* Mint show also something strange with this** and this*** prefixes,
need to check what is causing this, invalid code in bucket policy functions
or Mint or Client

nl5887 committed:
Strip extra wildcard from prefix

nl5887 committed:
Fix latest bucket policy related errors

vadmeste commented on cmd/s3-layer.go:
```diff
@@ -0,0 +1,588 @@
+	return ListObjectsV2Info{
+		IsTruncated: result.IsTruncated,
+		Prefixes: []string{
+			result.Prefix,
```
I think you want to fill Prefixes with result.CommonPrefixes elements instead.

vadmeste commented on cmd/s3-layer.go:
```diff
@@ -0,0 +1,588 @@
+	return ListObjectsInfo{
+		IsTruncated: result.IsTruncated,
+		NextMarker:  result.NextMarker,
+		Prefixes: []string{
```
I think you want to fill Prefixes with result.CommonPrefixes elements instead.

vadmeste commented on cmd/s3-layer.go:
```diff
@@ -0,0 +1,588 @@
+
+	defer object.Close()
+
+	object.Seek(startOffset, io.SeekStart)
```
Better to check errors returned from object.Seek()

vadmeste commented on cmd/s3-layer.go:
```diff
@@ -0,0 +1,588 @@
+func (l *s3Gateway) CompleteMultipartUpload(bucket string, object string, uploadID string, uploadedParts []completePart) (ObjectInfo, error) {
+	err := l.Client.CompleteMultipartUpload(bucket, object, uploadID, toMinioClientCompleteParts(uploadedParts))
+	if err != nil {
+		return ObjectInfo{}, nil
```
We should return an error here, not nil.

nl5887 committed:
Merge branch 'master' of https://github.com/minio/minio into feature-gateway-cache

nl5887 committed:
Return error instead of nil

nl5887 committed:
Return error from Seek

nl5887 committed:
Use CommonPrefix instead

nl5887 committed:
Fix linting error, omit 2nd value from range

nl5887 committed:
Fix linting error, should have comment

nl5887 committed:
Merge branch 'master' of https://github.com/minio/minio into feature-gateway-cache

nl5887 committed:
Use CommonPrefixes, fixes #4129

nl5887 committed:
Fix typo in comment

harshavardhana committed:
Merge branch 'master' into feature-gateway-cache

nl5887 committed:
Fix return wrong error on policy conflict, fixes #4127

nl5887 committed:
Merge branch 'feature-gateway-cache' of github.com:dutchcoders/minio into feature-gateway-cache

harshavardhana commented on cmd/s3-layer.go:
```diff
@@ -0,0 +1,597 @@
+
+// fromMinioClientListBucketPolicies converts from map[string]policy.BucketPolicy to []BucketAccessPolicy
+func fromMinioClientListBucketPolicies(bucket string, bucketPolicies map[string]policy.BucketPolicy) []BucketAccessPolicy {
+	policies := make([]BucketAccessPolicy, 0)
```
Can this be fixed https://github.com/golang/go/wiki/CodeReviewComments#declaring-empty-slices?

harshavardhana commented on cmd/s3-layer.go:
```diff
@@ -0,0 +1,597 @@
+
+// fromMinioClientListBucketPolicies converts from map[string]policy.BucketPolicy to []BucketAccessPolicy
+func fromMinioClientListBucketPolicies(bucket string, bucketPolicies map[string]policy.BucketPolicy) []BucketAccessPolicy {
+	policies := make([]BucketAccessPolicy, 0)
```
Can this be fixed? - https://github.com/golang/go/wiki/CodeReviewComments#declaring-empty-slices

nl5887 committed:
Update minio-go lib

nl5887 committed:
Implement the new core policy functions

nl5887 committed:
Merge branch 'master' of https://github.com/minio/minio into feature-gateway-cache

nl5887 committed:
Bump minio-go library

nl5887 committed:
Add SHA256Mismatch error

nl5887 committed:
Fix testing error with NotSupported
